src/modules/tools/reporter/Reporter.cpp

```diff
@@ -0,0 +1,147 @@
+*/
+
+#include "libs/Module.h"
+#include "libs/Kernel.h"
+#include "Reporter.h"
+#include "Gcode.h"
+#include "StreamOutput.h"
+#include "TemperatureControlPublicAccess.h"
+#include "TemperatureControl.h"
+#include "PlayerPublicAccess.h"
+#include "Player.h"
+#include "ExtruderPublicAccess.h"
+#include "PublicData.h"
+#include "Robot.h"
+#include "modules/robot/Conveyor.h"
+
+#include "version.h"
+#include "checksumm.h"
+#include "ConfigValue.h"
+#include "Config.h"
+
+#include <math.h>
+#include <vector>
+
+#define reporter_checksum          CHECKSUM("reporter")
+#define enable_checksum            CHECKSUM("enable")
+
+Reporter::Reporter(){}
+
+Reporter::~Reporter(){}
+
+void Reporter::on_module_loaded(){
+  // Exit if this module is not enabled
+  if ( !THEKERNEL->config->value( reporter_checksum, enable_checksum )->by_default(false)->as_bool() ) {
+      delete this;
+      return;
+  }
+
+  this->register_for_event(ON_GCODE_RECEIVED);
+}
+
+void Reporter::on_gcode_received(void *argument){
+    Gcode *gcode = static_cast<Gcode *>(argument);
+    if (gcode->has_m) {
+        if( gcode->m == 408 ) {
+            if (gcode->has_letter('S')) {
+                switch(static_cast<int>(gcode->get_value('S'))){
+                        case 0:
+                              // Beginning
+                              // Find the status
+                              //I=idle, P=printing from SD card, S=stopped (i.e. needs a reset),
+                              //C=running config file, A=paused, D=pausing, R=resuming, B=busy (running a macro)
+                              gcode->stream->printf("{\"status\":\"");
+                              if(THEKERNEL->conveyor->is_queue_empty()){
+                                gcode->stream->printf("I");
+                              }else{
+                                gcode->stream->printf("P");
+                              }
+                              if (THEKERNEL->is_halted()){
+                                gcode->stream->printf("S");
+                              }
+                              void *returned_data;
+                              if (PublicData::get_value( player_checksum, is_suspended_checksum, &returned_data )) {
+                                bool b = *static_cast<bool *>(returned_data);
+                                if(b){
+                                  gcode->stream->printf("A");
+                                }
+                              }
+                              gcode->stream->printf("\",\"heaters{{\":[");
+                              std::vector<struct pad_temperature> controllers;
+                              if (PublicData::get_value(temperature_control_checksum, poll_controls_checksum, &controllers)) {
+                                for (auto &c : controllers) {
+                                  if(&c == &controllers.back()){
```

Comment:
It's supposed to do the , correctly for the Json array but I copy and pasted poorly. I need to rethink this anyway as the heater array for M408 says "current heater temperatures, numbered as per the machine (typically, heater 0 is the bed)".